### PR TITLE
Add transmute range assumes for pointers too

### PR DESCRIPTION
**Pull request by saethlin**
r? @ghost

### CODE REVIEW

File: tests/codegen/intrinsics/transmute-niched.rs [hunk 1]
```diff
@@ -181,3 +181,31 @@ pub unsafe fn check_bool_to_ordering(x: bool) -> std::cmp::Ordering {
...
 }
+
+// CHECK-LABEL: @check_usize_to_ref(
+#[no_mangle]
+pub unsafe fn check_usize_to_ref(x: usize) -> &'static u32 {
+    // CHECK: %_0 = inttoptr i64 %x to ptr
+    // OPT: %0 = ptrtoint ptr %_0 to i64
+    // OPT: %1 = icmp uge i64 %0, 1
+    // OPT: call void @llvm.assume(i1 %1)
+    // DGB-NOT: icmp
```
Comment (the8472) [inline tests/codegen/intrinsics/transmute-niched.rs:192]: looks like a typo

**Comment by saethlin** [comment]
We only recognize the null pointer as an invalid state for references, so I'm trying the IR sequence that only asserts references are not null.
@bors try @rust-timer queue

**Comment by saethlin** [comment]
I'm closing this for the cleanliness of my own PR list.

But in general I think the transmute range assumes are a neat trick, but ultimately the wrong way to apply range information, especially for pointers.

**Pull request closed without merging**